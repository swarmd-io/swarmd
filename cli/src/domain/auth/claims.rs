use serde::{Deserialize, Serialize};
use serde_with::{serde_as, skip_serializing_none};
use std::{collections::HashMap, time::Duration};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum OneOrMany<T> {
    /// Single value
    One(T),
    /// Array of values
    Vec(Vec<T>),
}

impl<T> OneOrMany<T> {
    #[inline]
    fn is_empty(&self) -> bool {
        matches!(self, OneOrMany::Vec(v) if v.is_empty())
    }
}

impl<T> Default for OneOrMany<T> {
    #[inline]
    fn default() -> Self {
        Self::Vec(Vec::new())
    }
}

#[serde_as]
#[skip_serializing_none]
#[non_exhaustive]
#[derive(Debug, Serialize, Default, Deserialize)]
pub struct Claims<T> {
    #[serde_as(as = "Option<serde_with::DurationSeconds<f64>>")]
    pub exp: Option<Duration>,
    #[serde_as(as = "Option<serde_with::DurationSeconds<f64>>")]
    pub nbf: Option<Duration>,
    #[serde_as(as = "Option<serde_with::DurationSeconds<f64>>")]
    pub iat: Option<Duration>,

    pub iss: Option<String>,
    pub sub: Option<String>,
    #[serde(default, skip_serializing_if = "OneOrMany::is_empty")]
    pub aud: OneOrMany<String>,
    pub jti: Option<String>,

    #[serde(flatten)]
    pub extra: T,
}

// Should be in sync with JWT template used for CLI
#[derive(Debug, Serialize, Deserialize)]
pub struct ExtraClaims {
    clerk_user_id: String,
    cli: bool,
    current_org_clerk_id: Option<String>,
    pub default_org: String,
    org_slug: Option<String>,
    // Org_id -> Role
    orgs: HashMap<String, String>,
}
