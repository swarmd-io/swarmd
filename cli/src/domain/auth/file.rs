use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct AuthFile {
    pub token: String,
}
