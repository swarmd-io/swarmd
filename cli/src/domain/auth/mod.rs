use std::{
    fs::{remove_file, File},
    io::{BufReader, BufWriter},
};

use anyhow::{bail, Context};

mod file;
use base64::{
    alphabet::URL_SAFE,
    engine::{DecodePaddingMode, GeneralPurpose, GeneralPurposeConfig},
};
use console::{style, Emoji};
use file::AuthFile;

mod claims;
use claims::{Claims, ExtraClaims};

use super::Env;

pub struct AuthContext {
    token: String,
    pub claims: Claims<ExtraClaims>,
}

const ENGINE_CONFIG: GeneralPurposeConfig = GeneralPurposeConfig::new()
    .with_encode_padding(false)
    .with_decode_padding_mode(DecodePaddingMode::RequireNone)
    .with_decode_allow_trailing_bits(true);

const ENGINE: GeneralPurpose = GeneralPurpose::new(&URL_SAFE, ENGINE_CONFIG);

const AUTH_FILE: &str = "auth.json";

static WARNING: Emoji<'_, '_> = Emoji("⚠️ ", "");

impl AuthContext {
    pub fn token(&self) -> &String {
        &self.token
    }
    pub fn new_from_token(token: String) -> anyhow::Result<Self> {
        let mut parts = token.split('.');

        let _header = parts.next().context("Invalid token: no header")?;
        let mut payload = parts.next().context("Invalid token")?.as_bytes();
        let payload_r = base64::read::DecoderReader::new(&mut payload, &ENGINE);
        let claims: Claims<ExtraClaims> = serde_json::from_reader(payload_r)?;

        Ok(Self { token, claims })
    }

    pub fn save(self) -> anyhow::Result<()> {
        let base =
            crate::infrastructure::fs::base_directory().context("Couln't load base directory")?;

        let path = base.place_state_file(AUTH_FILE)?;
        let auth_file = File::create(path)?;
        let writer = BufWriter::new(auth_file);
        let auth = AuthFile { token: self.token };
        serde_json::to_writer_pretty(writer, &auth)?;
        Ok(())
    }

    /// Load AuthContext from environment by loading the associated token through the associated
    /// file.
    pub fn from_env() -> anyhow::Result<Option<Self>> {
        let base =
            crate::infrastructure::fs::base_directory().context("Couln't load base directory")?;

        let file_path = match base.find_state_file(AUTH_FILE) {
            None => return Ok(None),
            Some(file) => file,
        };

        let file_opened = File::open(file_path.clone()).context("Couldn't open file")?;
        let reader = BufReader::new(file_opened);
        let auth_file: AuthFile = match serde_json::from_reader(reader) {
            Err(err) => {
                tracing::error!("{err:?}");
                // If the file is corrupted, we delete it.
                remove_file(file_path).context("Coudln't delete auth file")?;
                return Ok(None);
            }
            Ok(elt) => elt,
        };

        Ok(Some(Self::new_from_token(auth_file.token)?))
    }

    pub fn auth_cached(env: &Env) -> anyhow::Result<Self> {
        if let Some(auth) = Self::from_env()? {
            Ok(auth)
        } else {
            env.println(format!(
                "{} {}You must be authentificated, run `swarmd login`!",
                style("").red().bold().dim(),
                WARNING
            ))?;
            bail!("You must be logged in.");
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_token_snapshot() {
        let token_test = "no_header.eyJjbGVya191c2VyX2lkIjoidXNlcl8yWG82c0dQU0JmMlVnS0hOeWRZMnBqVFNncXkiLCJjbGkiOnRydWUsImN1cnJlbnRfb3JnX2NsZXJrX2lkIjpudWxsLCJkZWZhdWx0X29yZyI6Im9yZ18yWG83TDRjbHZPSEp2a0hjSjZ1Y29JTE5JNkMiLCJleHAiOjE2OTk2OTk4ODgsImlhdCI6MTY5OTYxMzQ4OCwiaXNzIjoiaHR0cHM6Ly9ub3JtYWwta2lkLTM4LmNsZXJrLmFjY291bnRzLmRldiIsImp0aSI6ImQ2NTA1ZjkwYTk4OWUwMjg4ODE2IiwibmJmIjoxNjk5NjEzNDgzLCJvcmdfc2x1ZyI6bnVsbCwib3JncyI6eyJvcmdfMlhvN0w0Y2x2T0hKdmtIY0o2dWNvSUxOSTZDIjoiYWRtaW4ifSwic3ViIjoidXNlcl8yWG82c0dQU0JmMlVnS0hOeWRZMnBqVFNncXkiLCJ1c2VyX2lkIjpudWxsfQ.no_signin";

        let mut parts = token_test.split('.');

        let _header = parts.next().context("Invalid token: no header").unwrap();
        let mut payload = parts.next().context("Invalid token").unwrap().as_bytes();
        let payload_r = base64::read::DecoderReader::new(&mut payload, &ENGINE);
        let claims: Claims<ExtraClaims> = serde_json::from_reader(payload_r).unwrap();
        insta::assert_yaml_snapshot!(claims);
    }
}
