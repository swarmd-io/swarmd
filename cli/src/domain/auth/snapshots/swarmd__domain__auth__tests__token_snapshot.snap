---
source: cli/src/domain/auth/mod.rs
assertion_line: 118
expression: claims
---
exp: 1699699888
nbf: 1699613483
iat: 1699613488
iss: "https://normal-kid-38.clerk.accounts.dev"
sub: user_2Xo6sGPSBf2UgKHNydY2pjTSgqy
jti: d6505f90a989e0288816
clerk_user_id: user_2Xo6sGPSBf2UgKHNydY2pjTSgqy
cli: true
current_org_clerk_id: ~
default_org: org_2Xo7L4clvOHJvkHcJ6ucoILNI6C
org_slug: ~
orgs:
  org_2Xo7L4clvOHJvkHcJ6ucoILNI6C: admin

